src/hotspot/share/runtime/handshake.cpp

```diff
@@ -405,14 +405,14 @@ HandshakeState::HandshakeState(JavaThread* target) :
   _handshakee(target),
   _queue(),
   _lock(Monitor::leaf, "HandshakeState", Mutex::_allow_vm_block_flag, Monitor::_safepoint_check_never),
-  _active_handshaker()
+  _active_handshaker(),
+  _suspended(false),
+  _async_suspend_handshake(false)
 {
 }
 
 void HandshakeState::add_operation(HandshakeOperation* op) {
   // Adds are done lock free and so is arming.
-  // Calling this method with lock held is considered an error.
-  assert(!_lock.owned_by_self(), "Lock should not be held");
```

Comment:
I originally added this comment inside the "resolved" conversation and that kept
this comment from showing up. Let's try it as a new comment.

Doesn't that mean the comment on L415 needs updating?
Should it be something like:

// Synchronous adds are done lock free and so is arming, but some
// asynchronous adds are done when we already hold the lock.

I'm not sure about the "some asynchronous adds" part...
@dcubed-ojdk

Reply:
The mutex is unrelated to adds/inserts.
Adds/inserts to the queue can always be done regardless of which locks a thread may or may not have.
The reason for not allowing inserts while holding the HandshakeState lock was to eliminate that from the table, since it have other implications.
As @pchilano found we needed to reverse the order in should_process() to make it work (which I missed).

So now when we have looked at the case, it is okay to do it.

Meaning that you can add a handshake to yourself from another handshake regardless of it is sync or async.